File: PHP/WebsiteCode/LeechCode/Medium/Array/TakeKOfEachCharacterFromLeftAndRight.php
<?php  

// 105ms 24.66MB
class Solution {

    /**
     * @param string $s
     * @param int $k
     * @return int
     */
    function takeCharacters($s, $k) {
        // Check if k is 0; if so, no characters need to be taken
        if ($k == 0) {
            return 0;
        }
        
        // Count the total occurrences of each character in the string
        $charCount = array_count_values(str_split($s));
        $requiredChars = ['a' => 0, 'b' => 0, 'c' => 0];
        
        foreach (['a', 'b', 'c'] as $char) {
            // If any character count is less than k, return -1
            if (($charCount[$char] ?? 0) < $k) {
                return -1;
            }
            $requiredChars[$char] = $k;
        }

        // Sliding window to minimize characters we need to take
        $n = strlen($s);
        $left = 0;
        $right = 0;
        $currentCount = ['a' => 0, 'b' => 0, 'c' => 0];
        $maxLength = 0;

        while ($right < $n) {
            $currentCount[$s[$right]]++;
            $right++;

            // Ensure that the window does not exceed the allowed count
            while ($currentCount['a'] > $charCount['a'] - $k ||
                   $currentCount['b'] > $charCount['b'] - $k ||
                   $currentCount['c'] > $charCount['c'] - $k) {
                $currentCount[$s[$left]]--;
                $left++;
            }

            // Maximize the length of the valid subarray
            $maxLength = max($maxLength, $right - $left);
        }

        // Minimum minutes required is the total length minus the max valid window
        return $n - $maxLength;
    }
}

File: Python/WebsiteCode/LeetCode/Medium/Array/TakeKOfEachCharacterFromLeftAndRight.py
# 123ms 17.12MB
class Solution:
    def takeCharacters(self, s: str, k: int) -> int:
        freqs = [0] * 3
        n = len(s)
        
        for c in s:
            freqs[ord(c) - ord('a')] += 1
        
        i = 0
        j = 0
        if freqs[0] < k or freqs[1] < k or freqs[2] < k:
            return -1
        
        for j in range(n):
            freqs[ord(s[j]) - ord('a')] -= 1
            
            if freqs[0] < k or freqs[1] < k or freqs[2] < k:
                freqs[ord(s[i]) - ord('a')] += 1
                i += 1
            

        return n - (j - i + 1)
                
        
            

# 221ms 17.24MB
class Solution:
    def takeCharacters(self, s: str, k: int) -> int:
        count = [0] * 3
        n = len(s)

        # Count total occurrences
        for c in s:
            count[ord(c) - ord("a")] += 1

        # Check if we have enough characters
        for i in range(3):
            if count[i] < k:
                return -1

        window = [0] * 3
        left, max_window = 0, 0

        # Find the longest window that leaves k of each character outside
        for right in range(n):
            window[ord(s[right]) - ord("a")] += 1

            # Shrink window if we take too many characters
            while left <= right and (
                count[0] - window[0] < k
                or count[1] - window[1] < k
                or count[2] - window[2] < k
            ):
                window[ord(s[left]) - ord("a")] -= 1
                left += 1

            max_window = max(max_window, right - left + 1)

        return n - max_window